### PR TITLE
Use gitlab v4 api

### PR DESCRIPTION
**Pull request by sijis**
We should use version 4 of gitlab api. The api v3 will be removed in a future version later this year.

### CODE REVIEW

File: tests/utils/test_file_lookups.py [hunk 1]
```diff
@@ -50,58 +49,51 @@ def test_project_id_exception(mock_gitlab):
...
-    project_id = 30
-
-    mock_gitlab.Gitlab.return_value.getproject.return_value = {'id': project_id}
-
+    project_id = object
+    mock_gitlab.Gitlab.return_value.projects.get.return_value = object
     seeker = FileLookup()
-
-    assert seeker.project_id == project_id
+    assert seeker.project == project_id
```
Comment (e4r7hbug) [inline tests/utils/test_file_lookups.py]: This is testing equivalency. You should test that they are identical with `is`. The return value of `...projects.get()` should be identical to `self.project`.

File: tests/utils/test_file_lookups.py [hunk 2]
```diff
@@ -50,58 +49,51 @@ def test_project_id_exception(mock_gitlab):
...
 def test_json(gitlab):
     """Check _json_ method."""
     my_git = FileLookup()
 
-    my_git.server.getfile.return_value = {'content': base64.b64encode(TEST_JSON_BYTES)}
-
-    result = my_git.json()
-    assert isinstance(result, dict)
-    assert result['ship'] == 'pirate'
+    with mock.patch.object(FileLookup, 'get', return_value=TEST_JSON) as mock_method:
```
Comment (e4r7hbug) [inline tests/utils/test_file_lookups.py]: Use decorator and `assert_called_with()` like above.

File: tests/utils/test_file_lookups.py [hunk 3]
```diff
@@ -50,58 +49,51 @@ def test_project_id_exception(mock_gitlab):
...
 @mock.patch('foremast.utils.lookups.gitlab')
 def test_invalid_json(gitlab):
     """Check invalid JSON."""
     my_git = FileLookup()
 
-    my_git.server.getfile.return_value = {'content': base64.b64encode(TEST_JSON_BYTES + b'}')}
-
-    with pytest.raises(SystemExit):
-        my_git.json()
+    with mock.patch.object(FileLookup, 'get', return_value=TEST_JSON + '}') as mock_method:
```
Comment (e4r7hbug) [inline tests/utils/test_file_lookups.py]: Use decorator and `assert_called_with()` like above.

File: tests/utils/test_file_lookups.py [hunk 4]
```diff
@@ -50,58 +49,51 @@ def test_project_id_exception(mock_gitlab):
...
+def test_filelookup_attr():
     """Make sure Git related attributes are missing when runway is specified."""
     my_git = FileLookup(runway_dir='/poop_deck')
 
-    assert getattr(my_git, 'git_short') is None
-    assert getattr(my_git, 'server') is None
-    assert getattr(my_git, 'project_id') is None
+    assert my_git.git_short == ''
+    assert my_git.server is None
+    assert my_git.project_id == ''
```
Comment (e4r7hbug) [inline tests/utils/test_file_lookups.py]: `project_id` should not exist. We now care about `project`.

File: src/foremast/utils/lookups.py [hunk 5]
```diff
@@ -103,15 +114,14 @@ def get_gitlab_project_id(self):
...
-        project = self.server.getproject(self.git_short)
+        self.server = gitlab.Gitlab(GIT_URL, private_token=GITLAB_TOKEN, api_version=4)
+        project = self.server.projects.get(self.git_short)
 
         if not project:
             raise GitLabApiError('Could not get Project "{0}" from GitLab API.'.format(self.git_short))
 
-        self.project_id = project['id']
-        return self.project_id
+        self.project = project
```
[the remark this answers is not in the document]
Comment (sijis) [reply src/foremast/utils/lookups.py]: Removing all references to `project_id`. its not needed for anything.

File: tests/test_ami_lookup.py [hunk 6]
```diff
@@ -21,15 +21,37 @@
...
     sample_json = json.dumps(sample_dict)
-    gitlab.return_value.getfile.return_value = {'content': base64.b64encode(sample_json.encode())}
-    assert ami_lookup(name='base_fedora').startswith('ami-xxxx')
-    assert ami_lookup(region='us-west-2').startswith('ami-xxxx')
+    ami_file.return_value = sample_json
+    assert ami_lookup(name='base_fedora') == 'ami-xxxx'
+    assert ami_lookup(region='us-west-2') == 'ami-yyyy'
+
+
+@mock.patch('foremast.utils.lookups.AMI_JSON_URL', return_value=True)
```
[the remark this answers is not in the document]
Comment (sijis) [reply tests/test_ami_lookup.py]: Fixed.

File: tests/test_ami_lookup.py [hunk 7]
```diff
@@ -21,15 +21,37 @@
 from foremast.utils import ami_lookup
 
 
-@mock.patch('foremast.utils.lookups.GITLAB_TOKEN')
-@mock.patch('foremast.utils.lookups.gitlab.Gitlab')
-def test_ami_lookup(gitlab, token):
+@mock.patch('foremast.utils.lookups.GITLAB_TOKEN', return_value=True)
```
[the remark this answers is not in the document]
Comment (sijis) [reply tests/test_ami_lookup.py]: fixed.

File: tests/utils/test_file_lookups.py (hunk 4 above)
Comment (sijis) [reply tests/utils/test_file_lookups.py]: Removed.

File: tests/utils/test_file_lookups.py (hunk 1 above)
Comment (sijis) [reply tests/utils/test_file_lookups.py]: Fixed.

File: tests/utils/test_file_lookups.py [hunk 8]
```diff
@@ -50,58 +49,51 @@ def test_project_id_exception(mock_gitlab):
 @mock.patch('foremast.utils.lookups.gitlab')
 def test_project_id_success(mock_gitlab):
     """Check resolving GitLab Project ID is successful."""
-    project_id = 30
-
-    mock_gitlab.Gitlab.return_value.getproject.return_value = {'id': project_id}
-
+    project_id = object
+    mock_gitlab.Gitlab.return_value.projects.get.return_value = object
```
[the remark this answers is not in the document]
Comment (sijis) [reply tests/utils/test_file_lookups.py]: I think i've addressed this with `is object`

File: src/foremast/utils/lookups.py [hunk 9]
```diff
@@ -43,21 +43,12 @@ def ami_lookup(region='us-east-1', name='tomcat8'):
 
     """
     if AMI_JSON_URL:
-        LOG.info("Getting AMI from %s", AMI_JSON_URL)
-        response = requests.get(AMI_JSON_URL)
-        assert response.ok, "Error getting ami info from {}".format(AMI_JSON_URL)
-
-        ami_dict = response.json()
+        ami_dict = _get_ami_json(AMI_JSON_URL)
         LOG.debug('Lookup AMI table: %s', ami_dict)
```
[the remark this answers is not in the document]
Comment (sijis) [reply src/foremast/utils/lookups.py]: Good point. Moved.

File: tests/utils/test_file_lookups.py (hunk 2 above)
Comment (sijis) [reply tests/utils/test_file_lookups.py]: Fixed. Definitely makes the tests easier to read.

File: tests/test_ami_lookup.py [hunk 10]
```diff
@@ -21,15 +21,37 @@
...
-    assert ami_lookup(name='base_fedora').startswith('ami-xxxx')
-    assert ami_lookup(region='us-west-2').startswith('ami-xxxx')
+    ami_file.return_value = sample_json
+    assert ami_lookup(name='base_fedora') == 'ami-xxxx'
+    assert ami_lookup(region='us-west-2') == 'ami-yyyy'
+
+
+@mock.patch('foremast.utils.lookups.AMI_JSON_URL', return_value=True)
+@mock.patch('foremast.utils.lookups._get_ami_json')
+def test_json_lookup(ami_file, json_url):
```
[the remark this answers is not in the document]
Comment (sijis) [reply tests/test_ami_lookup.py]: renamed.

**Comment by sijis** [comment]
Rebased branch too from latest changes from `master`.

File: src/foremast/utils/lookups.py [hunk 11]
```diff
@@ -69,6 +60,26 @@ def ami_lookup(region='us-east-1', name='tomcat8'):
     return ami_id
 
 
+def _get_ami_file(region='us-east-1'):
+    """Helper function to get file from Gitlab."""
+    LOG.info("Getting AMI from Gitlab")
+    server = gitlab.Gitlab(GIT_URL, private_token=GITLAB_TOKEN, api_version=4)
```
[the remark this answers is not in the document]
Comment (sijis) [reply src/foremast/utils/lookups.py]: ohh. maybe!

File: tests/test_ami_lookup.py [hunk 12]
```diff
@@ -21,15 +21,37 @@
...
+            'tomcat8': 'ami-yyyy',
+        }
+    }
+    ami_file.return_value = sample_dict
+    assert ami_lookup(region='us-east-1', name='base_fedora') == 'ami-xxxx'
+    assert ami_lookup(region='us-west-2', name='tomcat8') == 'ami-yyyy'
+
+
+def test_no_external_lookup():
+    """AMI lookup should contact GitLab for JSON table and resolve."""
```
[the remark this answers is not in the document]
Comment (sijis) [reply tests/test_ami_lookup.py]: Fixed.

File: src/foremast/utils/lookups.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (sijis) [reply src/foremast/utils/lookups.py]: Refactored.

File: src/foremast/utils/lookups.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (e4r7hbug) [reply src/foremast/utils/lookups.py]: We want `self.project` to be initialized in the `__init__`.

File: src/foremast/utils/lookups.py [hunk 13]
```diff
@@ -43,23 +43,12 @@ def ami_lookup(region='us-east-1', name='tomcat8'):
...
-        LOG.info("Getting AMI from %s", AMI_JSON_URL)
-        response = requests.get(AMI_JSON_URL)
-        assert response.ok, "Error getting ami info from {}".format(AMI_JSON_URL)
-
-        ami_dict = response.json()
-        LOG.debug('Lookup AMI table: %s', ami_dict)
+        ami_dict = _get_ami_json(AMI_JSON_URL)
         ami_id = ami_dict[region][name]
     elif GITLAB_TOKEN:
         # TODO: Remove GitLab repository in favour of JSON URL option.
```
Comment (e4r7hbug) [inline src/foremast/utils/lookups.py]: Move TODO into function.

File: tests/utils/test_file_lookups.py (hunk 4 above)
Comment (e4r7hbug) [reply tests/utils/test_file_lookups.py]: Did not want it removed, should be verifying `self.project` is set to something expected.

File: src/foremast/utils/lookups.py [hunk 14]
```diff
@@ -69,6 +58,26 @@ def ami_lookup(region='us-east-1', name='tomcat8'):
...
+    """Get file from Gitlab."""
+    LOG.info("Getting AMI from Gitlab")
+    lookup = FileLookup(git_short='devops/ansible')
+    filename = 'scripts/{0}.json'.format(region)
+    ami_contents = lookup.remote_file(filename=filename, branch='master')
+    LOG.debug('AMI file contents in %s: %s', filename, ami_contents)
+    return ami_contents
+
+
+def _get_ami_json(json_url):
```
Comment (e4r7hbug) [inline src/foremast/utils/lookups.py]: Rename to `_get_ami_dict`.

File: src/foremast/utils/lookups.py [hunk 15]
```diff
@@ -83,16 +92,16 @@ class FileLookup():
     """
 
     def __init__(self, git_short='', runway_dir=''):
+        """Init."""
```
Comment (e4r7hbug) [inline src/foremast/utils/lookups.py]: We are standardizing on docstrings on the Class. No need to add to `__init__`.

File: src/foremast/utils/lookups.py (hunk 13 above)
Comment (sijis) [reply src/foremast/utils/lookups.py]: Made it a deprecation warning.

File: src/foremast/utils/lookups.py (hunk 15 above)
Comment (sijis) [reply src/foremast/utils/lookups.py]: pydoc is yelling at me. :)

File: src/foremast/utils/lookups.py (hunk 14 above)
Comment (sijis) [reply src/foremast/utils/lookups.py]: Renamed.

Comment (sijis) [reply src/foremast/utils/lookups.py]: I just made sense to do it.

File: src/foremast/configs/prepare_configs.py [hunk 16]
```diff
@@ -37,7 +37,8 @@ def process_git_configs(git_short=''):
     LOG.info('Processing application.json files from GitLab "%s".', git_short)
     file_lookup = FileLookup(git_short=git_short)
     app_configs = process_configs(file_lookup, 'runway/application-master-{env}.json', 'runway/pipeline.json')
-    config_commit = file_lookup.server.getbranch(file_lookup.project_id, 'master')['commit']['id']
+    commit_obj = file_lookup.project.commits.get('master')
+    config_commit = commit_obj.attributes.get('id')
```
[the remark this answers is not in the document]
Comment (sijis) [reply src/foremast/configs/prepare_configs.py]: Makes sense. fixed.

File: src/foremast/utils/lookups.py (hunk 15 above)
Comment (sijis) [reply src/foremast/utils/lookups.py]: Fixed this too.

File: src/foremast/utils/lookups.py [hunk 17]
```diff
@@ -69,6 +60,26 @@ def ami_lookup(region='us-east-1', name='tomcat8'):
...
+
+    ami_contents = b64decode(ami_blob.decode()).decode()
+    return ami_contents
+
+
+def _get_ami_json(json_url):
+    """Helper function to get ami from a web url."""
+    LOG.info("Getting AMI from %s", json_url)
+    response = requests.get(json_url)
+    assert response.ok, "Error getting ami info from {}".format(json_url)
```
[the remark this answers is not in the document]
Comment (sijis) [reply src/foremast/utils/lookups.py]: I just wanted to leave things as they were and not refactor.

**Pull request merged**